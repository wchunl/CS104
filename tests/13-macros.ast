ROOT "" (0.0.0)
|   TOK_FUNCTION "" (6.22.1)
|   |   TOK_TYPE_ID "" (6.22.1)
|   |   |   TOK_VOID "void" (6.22.1)
|   |   |   TOK_IDENT "fail" (6.22.6)
|   |   TOK_PARAM "(" (6.22.11)
|   |   |   TOK_TYPE_ID "" (6.22.12)
|   |   |   |   TOK_STRING "string" (6.22.12)
|   |   |   |   TOK_IDENT "expr" (6.22.19)
|   |   |   TOK_TYPE_ID "" (6.22.25)
|   |   |   |   TOK_STRING "string" (6.22.25)
|   |   |   |   TOK_IDENT "file" (6.22.32)
|   |   |   TOK_TYPE_ID "" (6.22.38)
|   |   |   |   TOK_INT "int" (6.22.38)
|   |   |   |   TOK_IDENT "line" (6.22.42)
|   TOK_FUNCTION "" (6.24.1)
|   |   TOK_TYPE_ID "" (6.24.1)
|   |   |   TOK_VOID "void" (6.24.1)
|   |   |   TOK_IDENT "putchr" (6.24.6)
|   |   TOK_PARAM "(" (6.24.13)
|   |   |   TOK_TYPE_ID "" (6.24.14)
|   |   |   |   TOK_INT "int" (6.24.14)
|   |   |   |   TOK_IDENT "chr" (6.24.18)
|   TOK_FUNCTION "" (6.25.1)
|   |   TOK_TYPE_ID "" (6.25.1)
|   |   |   TOK_VOID "void" (6.25.1)
|   |   |   TOK_IDENT "putint" (6.25.6)
|   |   TOK_PARAM "(" (6.25.13)
|   |   |   TOK_TYPE_ID "" (6.25.14)
|   |   |   |   TOK_INT "int" (6.25.14)
|   |   |   |   TOK_IDENT "num" (6.25.18)
|   TOK_FUNCTION "" (6.26.1)
|   |   TOK_TYPE_ID "" (6.26.1)
|   |   |   TOK_VOID "void" (6.26.1)
|   |   |   TOK_IDENT "putstr" (6.26.6)
|   |   TOK_PARAM "(" (6.26.13)
|   |   |   TOK_TYPE_ID "" (6.26.14)
|   |   |   |   TOK_STRING "string" (6.26.14)
|   |   |   |   TOK_IDENT "str" (6.26.21)
|   TOK_FUNCTION "" (6.28.1)
|   |   TOK_TYPE_ID "" (6.28.1)
|   |   |   TOK_INT "int" (6.28.1)
|   |   |   TOK_IDENT "getchr" (6.28.5)
|   |   TOK_PARAM "(" (6.28.11)
|   TOK_FUNCTION "" (6.29.1)
|   |   TOK_TYPE_ID "" (6.29.1)
|   |   |   TOK_STRING "string" (6.29.1)
|   |   |   TOK_IDENT "getstr" (6.29.8)
|   |   TOK_PARAM "(" (6.29.14)
|   TOK_FUNCTION "" (6.30.1)
|   |   TOK_TYPE_ID "" (6.30.1)
|   |   |   TOK_STRING "string" (6.30.1)
|   |   |   TOK_IDENT "getln" (6.30.8)
|   |   TOK_PARAM "(" (6.30.13)
|   TOK_FUNCTION "" (7.5.1)
|   |   TOK_TYPE_ID "" (7.5.1)
|   |   |   TOK_INT "int" (7.5.1)
|   |   |   TOK_IDENT "main" (7.5.5)
|   |   TOK_PARAM "(" (7.5.10)
|   |   |   TOK_TYPE_ID "" (7.5.11)
|   |   |   |   TOK_INT "int" (7.5.11)
|   |   |   |   TOK_IDENT "argc" (7.5.15)
|   |   |   TOK_TYPE_ID "" (7.5.21)
|   |   |   |   TOK_ARRAY "array" (7.5.21)
|   |   |   |   |   TOK_TYPE_ID "string" (7.5.27)
|   |   |   |   TOK_IDENT "argv" (7.5.35)
|   |   TOK_BLOCK "{" (7.5.41)
|   |   |   TOK_CALL "(" (7.6.11)
|   |   |   |   TOK_IDENT "putstr" (7.6.4)
|   |   |   |   TOK_INDEX "[" (7.6.16)
|   |   |   |   |   TOK_IDENT "argv" (7.6.12)
|   |   |   |   |   TOK_INTCON "0" (7.6.17)
|   |   |   |   TOK_CALL "(" (7.7.11)
|   |   |   |   |   TOK_IDENT "putstr" (7.7.4)
|   |   |   |   |   TOK_STRINGCON "": "" (7.7.12)
|   |   |   |   TOK_CALL "(" (7.8.11)
|   |   |   |   |   TOK_IDENT "putstr" (7.8.4)
|   |   |   |   |   TOK_STRINGCON ""13-macros.oc"" (7.8.12)
|   |   |   |   TOK_CALL "(" (7.9.11)
|   |   |   |   |   TOK_IDENT "putchr" (7.9.4)
|   |   |   |   |   TOK_CHARCON "'['" (7.9.12)
|   |   |   |   TOK_CALL "(" (7.10.11)
|   |   |   |   |   TOK_IDENT "putint" (7.10.4)
|   |   |   |   |   TOK_INTCON "10" (7.10.12)
|   |   |   |   TOK_CALL "(" (7.11.11)
|   |   |   |   |   TOK_IDENT "putstr" (7.11.4)
|   |   |   |   |   TOK_STRINGCON ""] compiled "" (7.11.12)
|   |   |   |   TOK_CALL "(" (7.12.11)
|   |   |   |   |   TOK_IDENT "putstr" (7.12.4)
|   |   |   |   |   TOK_STRINGCON ""May 24 2019"" (7.12.12)
|   |   |   |   TOK_CALL "(" (7.13.11)
|   |   |   |   |   TOK_IDENT "putstr" (7.13.4)
|   |   |   |   |   TOK_STRINGCON "" "" (7.13.12)
|   |   |   |   TOK_CALL "(" (7.14.11)
|   |   |   |   |   TOK_IDENT "putstr" (7.14.4)
|   |   |   |   |   TOK_STRINGCON ""00:37:42"" (7.14.12)
|   |   |   |   TOK_CALL "(" (7.15.11)
|   |   |   |   |   TOK_IDENT "putstr" (7.15.4)
|   |   |   |   |   TOK_STRINGCON ""\n"" (7.15.12)
|   |   |   |   TOK_RETURN "return" (7.16.4)
|   |   |   |   |   TOK_INTCON "0" (7.16.11)
