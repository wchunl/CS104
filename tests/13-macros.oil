fail:     .function
          .param ptr expr
          .param ptr file
          .param int line
          .end
putchr:   .function
          .param int chr
          .end
putint:   .function
          .param int num
          .end
putstr:   .function
          .param ptr str
          .end
getchr:   .function int
          .end
getstr:   .function
          .end
getln:    .function
          .end
main:     .function int
          .param int argc
          .param ptr argv
          $t0:p = argv[0 * :p]
          putstr($t0:i)
          putstr(": ")
          putstr("13-macros.oc")
          putchr('[')
          putint(10)
          putstr("] compiled ")
          putstr("Jun  6 2019")
          putstr(" ")
          putstr("19:07:00")
          putstr("\n")
          return 0
          return 
          .end
