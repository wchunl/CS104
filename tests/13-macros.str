string_set[   0]:    4706206694200543916 0x1b63c08->","
                   16579916199511856337 0x1b65098->"getln"
string_set[   1]:    2297668033614959926 0x1b65c38->"0"
                    3701972582333163920 0x1b65508->"argc"
string_set[   3]:   11868120863286193964 0x1b652b8->"main"
                    9611428142580219387 0x1b5f608->"fail"
string_set[   4]:      21606623943142546 0x1b64008->"{"
                    4017852018407011111 0x1b63cf8->"file"
string_set[   6]:   13686721134187162920 0x1b66bb8->"}"
string_set[   8]:    9386170080689500517 0x1b63ad8->"expr"
string_set[   9]:   14526079891305879776 0x1b658d8->">"
string_set[  11]:   17138325012746839705 0x1b668e8->""\n""
string_set[  12]:   15816506932648033945 0x1b63968->"("
string_set[  16]:   18335579659859433076 0x1b65e78->"": ""
string_set[  17]:   18413295556186398326 0x1b640f8->"putint"
string_set[  18]:    7211130878751977441 0x1b64228->"putchr"
                   12679825168387489523 0x1b64708->"num"
string_set[  19]:   14643508047410943861 0x1b65948->"argv"
string_set[  21]:   13581035385482130682 0x1b65b98->"["
string_set[  22]:   11369585928979698755 0x1b65cf8->""13-macros.oc""
string_set[  23]:   12051031859636400188 0x1b65cd8->"]"
                   11166701219533395393 0x1b64e58->"getstr"
                   12868445110721718657 0x1b63d18->")"
string_set[  28]:   14635384904416252381 0x1b66a88->"return"
string_set[  29]:    2594088158685378223 0x1b66078->"10"
                   12291392672316406109 0x1b63ea8->"line"
string_set[  30]:    2073966743809279764 0x1b66788->""14:51:29""
string_set[  31]:   14742483560488554491 0x1b664c8->""May 13 2019""
                    6142509188972423790 0x1b5f5e8->""
                   13348923631108115385 0x1b644e8->"putstr"
string_set[  33]:    1583576014053361350 0x1b64888->"str"
string_set[  38]:     814550533774441920 0x1b66348->""] compiled ""
                   12597418120160561823 0x1b63b48->"int"
string_set[  39]:    7296690267450279874 0x1b64368->"chr"
string_set[  40]:   11911796782241914524 0x1b5f218->"void"
string_set[  42]:    5344662657899890615 0x1b63ec8->";"
string_set[  43]:    6814028903552918164 0x1b65748->"array"
string_set[  44]:   10010194838115560012 0x1b66628->"" ""
                   13497578857207247412 0x1b63a08->"string"
string_set[  45]:    1396094175042499165 0x1b657e8->"<"
string_set[  46]:    3012229484421721799 0x1b65f18->"'['"
                    9201800968376702500 0x1b64c38->"getchr"
load_factor = 0.894
bucket_count = 47
max_bucket_size = 3
