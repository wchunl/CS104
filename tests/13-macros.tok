#  1 "13-macros.oc"
#  1 "<built-in>"
#  1 "<command-line>"
#  1 "13-macros.oc"
#  1 "oclib.h"
# 22 "oclib.h"
   6 22.00001 261 TOK_VOID      void
   6 22.00006 282 TOK_IDENT     fail
   6 22.00011  40 '('           (
   6 22.00012 263 TOK_STRING    string
   6 22.00019 282 TOK_IDENT     expr
   6 22.00023  44 ','           ,
   6 22.00025 263 TOK_STRING    string
   6 22.00032 282 TOK_IDENT     file
   6 22.00036  44 ','           ,
   6 22.00038 262 TOK_INT       int
   6 22.00042 282 TOK_IDENT     line
   6 22.00046  41 ')'           )
   6 22.00047  59 ';'           ;
   6 24.00001 261 TOK_VOID      void
   6 24.00006 282 TOK_IDENT     putchr
   6 24.00013  40 '('           (
   6 24.00014 262 TOK_INT       int
   6 24.00018 282 TOK_IDENT     chr
   6 24.00021  41 ')'           )
   6 24.00022  59 ';'           ;
   6 25.00001 261 TOK_VOID      void
   6 25.00006 282 TOK_IDENT     putint
   6 25.00013  40 '('           (
   6 25.00014 262 TOK_INT       int
   6 25.00018 282 TOK_IDENT     num
   6 25.00021  41 ')'           )
   6 25.00022  59 ';'           ;
   6 26.00001 261 TOK_VOID      void
   6 26.00006 282 TOK_IDENT     putstr
   6 26.00013  40 '('           (
   6 26.00014 263 TOK_STRING    string
   6 26.00021 282 TOK_IDENT     str
   6 26.00024  41 ')'           )
   6 26.00025  59 ';'           ;
   6 28.00001 262 TOK_INT       int
   6 28.00005 282 TOK_IDENT     getchr
   6 28.00011  40 '('           (
   6 28.00012  41 ')'           )
   6 28.00013  59 ';'           ;
   6 29.00001 263 TOK_STRING    string
   6 29.00008 282 TOK_IDENT     getstr
   6 29.00014  40 '('           (
   6 29.00015  41 ')'           )
   6 29.00016  59 ';'           ;
   6 30.00001 263 TOK_STRING    string
   6 30.00008 282 TOK_IDENT     getln
   6 30.00013  40 '('           (
   6 30.00014  41 ')'           )
   6 30.00015  59 ';'           ;
#  4 "13-macros.oc"
   7 5.00001 262 TOK_INT       int
   7 5.00005 282 TOK_IDENT     main
   7 5.00010  40 '('           (
   7 5.00011 262 TOK_INT       int
   7 5.00015 282 TOK_IDENT     argc
   7 5.00019  44 ','           ,
   7 5.00021 271 TOK_ARRAY     array
   7 5.00026  60 '<'           <
   7 5.00027 263 TOK_STRING    string
   7 5.00033  62 '>'           >
   7 5.00035 282 TOK_IDENT     argv
   7 5.00039  41 ')'           )
   7 5.00041 123 '{'           {
   7 6.00004 282 TOK_IDENT     putstr
   7 6.00011  40 '('           (
   7 6.00012 282 TOK_IDENT     argv
   7 6.00016  91 '['           [
   7 6.00017 283 TOK_INTCON    0
   7 6.00018  93 ']'           ]
   7 6.00019  41 ')'           )
   7 6.00020  59 ';'           ;
   7 7.00004 282 TOK_IDENT     putstr
   7 7.00011  40 '('           (
   7 7.00012 285 TOK_STRINGCON ": "
   7 7.00016  41 ')'           )
   7 7.00017  59 ';'           ;
   7 8.00004 282 TOK_IDENT     putstr
   7 8.00011  40 '('           (
   7 8.00012 285 TOK_STRINGCON "13-macros.oc"
   7 8.00026  41 ')'           )
   7 8.00027  59 ';'           ;
   7 9.00004 282 TOK_IDENT     putchr
   7 9.00011  40 '('           (
   7 9.00012 284 TOK_CHARCON   '['
   7 9.00015  41 ')'           )
   7 9.00016  59 ';'           ;
   7 10.00004 282 TOK_IDENT     putint
   7 10.00011  40 '('           (
   7 10.00012 283 TOK_INTCON    10
   7 10.00014  41 ')'           )
   7 10.00015  59 ';'           ;
   7 11.00004 282 TOK_IDENT     putstr
   7 11.00011  40 '('           (
   7 11.00012 285 TOK_STRINGCON "] compiled "
   7 11.00025  41 ')'           )
   7 11.00026  59 ';'           ;
   7 12.00004 282 TOK_IDENT     putstr
   7 12.00011  40 '('           (
   7 12.00012 285 TOK_STRINGCON "Jun  6 2019"
   7 12.00025  41 ')'           )
   7 12.00026  59 ';'           ;
   7 13.00004 282 TOK_IDENT     putstr
   7 13.00011  40 '('           (
   7 13.00012 285 TOK_STRINGCON " "
   7 13.00015  41 ')'           )
   7 13.00016  59 ';'           ;
   7 14.00004 282 TOK_IDENT     putstr
   7 14.00011  40 '('           (
   7 14.00012 285 TOK_STRINGCON "19:07:00"
   7 14.00022  41 ')'           )
   7 14.00023  59 ';'           ;
   7 15.00004 282 TOK_IDENT     putstr
   7 15.00011  40 '('           (
   7 15.00012 285 TOK_STRINGCON "\n"
   7 15.00016  41 ')'           )
   7 15.00017  59 ';'           ;
   7 16.00004 268 TOK_RETURN    return
   7 16.00011 283 TOK_INTCON    0
   7 16.00012  59 ';'           ;
   7 17.00001 125 '}'           }
