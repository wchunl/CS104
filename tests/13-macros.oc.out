; 0.  1: # 1 "13-macros.oc"; 1.  0: 
; 1.  1: # 1 "<built-in>"
; 2.  1: # 1 "<command-line>"
; 3.  1: # 1 "13-macros.oc"
; 4.  1: 
; 4.  2: 
; 4.  3: # 1 "oclib.h" 1
; 5.  1: # 22 "oclib.h"
; 6. 22: void fail (string expr, string file, int line);
; 6. 23: 
; 6. 24: void putchr (int chr);
; 6. 25: void putint (int num);
; 6. 26: void putstr (string str);
; 6. 27: 
; 6. 28: int getchr();
; 6. 29: string getstr();
; 6. 30: string getln();
; 6. 31: # 4 "13-macros.oc" 2
; 7.  4: 
; 7.  5: int main (int argc, array<string> argv) {
; 7.  6:    putstr (argv[0]);
; 7.  7:    putstr (": ");
; 7.  8:    putstr ("13-macros.oc");
; 7.  9:    putchr ('[');
; 7. 10:    putint (10);
; 7. 11:    putstr ("] compiled ");
; 7. 12:    putstr ("May 13 2019");
; 7. 13:    putstr (" ");
; 7. 14:    putstr ("14:51:29");
; 7. 15:    putstr ("\n");
; 7. 16:    return 0;
; 7. 17: }
hello